### PR TITLE
Add verify_unplug() for controller

### PR DESCRIPTION
**Pull request by peixiu**
Add verify_unplug() after unplug controller.

id: 1549937

Signed-off-by: Peixiu Hou <phou@redhat.com>

### CODE REVIEW

**Comment by peixiu** [comment]
(1/2) smp_32.16384m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win7.x86_64.sp1.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_repetition.one_pci: PASS (888.88 s)
 (2/2) smp_32.16384m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win7.x86_64.sp1.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_repetition.multi_pci: PASS (1742.03 s)

**Comment by peixiu** [comment]
@CongLi @aliang123 could you help to review this patch? thanks a lot~

**Comment by CongLi** [comment]
please add case in the commit log, like 'block_hotplug: ****'

File: qemu/tests/block_hotplug.py [hunk 1]
```diff
@@ -228,6 +230,8 @@ def unplug_device(vm, get_disk_cmd, device):
                     test.fail("Failed to unplug disks '%s'" % device.get_param("id"))
                 device_list.remove(device)
             controller.unplug(vm.monitor)
+            time.sleep(5)
```
Comment (aliang123) [inline qemu/tests/block_hotplug.py]: Defined variant "pause" can be used here, like time.sleep(pause).

File: qemu/tests/block_hotplug.py [hunk 2]
```diff
@@ -64,6 +65,7 @@ def unplug_device(vm, get_disk_cmd, device):
         """
         disks_before_unplug = find_disk(vm, get_disk_cmd)
         device.unplug(vm.monitor)
+        time.sleep(5)
```
Comment (aliang123) [inline qemu/tests/block_hotplug.py]: Suggestion 1: Add comment to explain why sleep 5s here.
Suggestion 2: If not sure rescan time for different windows, suggest to execute recan manually in def find_disk() before "wmic diskdrive get index" executed, like rescan_cmd = params.get("disk_rescan_cmd") session.cmd(rescan_cmd).

**Comment by peixiu** [comment]
Tried test with the latest code:
 (1/2) smp_32.16384m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win7.x86_64.sp1.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_repetition.one_pci: PASS (1186.90 s)
 (2/2) smp_32.16384m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win7.x86_64.sp1.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_repetition.multi_pci: PASS (3465.45 s)

Also tried one block_hotplug..with_block_resize case, it can be passed.
 (1/1) smp_32.16384m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win7.x86_64.sp1.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_block_resize.one_pci: PASS (222.24 s)

File: qemu/tests/block_hotplug.py (hunk 2 above)
Comment (peixiu) [reply qemu/tests/block_hotplug.py]: thanks aliang, updated~

File: qemu/tests/block_hotplug.py (hunk 1 above)
Comment (peixiu) [reply qemu/tests/block_hotplug.py]: updated, thanks

File: qemu/tests/block_hotplug.py [hunk 3]
```diff
@@ -37,15 +38,17 @@ def find_disk(vm, cmd):
         """
         Find all disks in guest.
         """
+        session = vm.wait_for_login(timeout=timeout)
         if params.get("os_type") == "linux":
             pattern = params.get("get_disk_pattern", "^/dev/vd[a-z]*$")
         elif params.get("os_type") == "windows":
+            disk_rescan_cmd = params.get("disk_rescan_cmd")
```
Comment (CongLi) [inline qemu/tests/block_hotplug.py]: could you explain why rescan here?

File: qemu/tests/block_hotplug.py [hunk 4]
```diff
@@ -228,6 +231,8 @@ def unplug_device(vm, get_disk_cmd, device):
                     test.fail("Failed to unplug disks '%s'" % device.get_param("id"))
                 device_list.remove(device)
             controller.unplug(vm.monitor)
+            time.sleep(pause)
+            controller.verify_unplug("", vm.monitor)
```
Comment (CongLi) [inline qemu/tests/block_hotplug.py]: could you please use wait_for() to check if unplug successfully?

File: qemu/tests/block_hotplug.py (hunk 3 above)
Comment (peixiu) [reply qemu/tests/block_hotplug.py]: Hi coli, please refer to the aliang's last comment, she suggest to replace the time.sleep(5) with rescan operation, due to not sue if 5 seconds is enough for guest rescan, so add a force rescan here.

File: qemu/tests/block_hotplug.py (hunk 4 above)
Comment (peixiu) [reply qemu/tests/block_hotplug.py]: Hi coli, the controller.verify_unplug("", vm.monitor) will verify if the controller has been deleted, if not be deleted will report False, why need use wait_for() to check it again?

Thanks~

Comment (CongLi) [reply qemu/tests/block_hotplug.py]: wait_for() could combine time.sleep() and verify_unplug together, and which is more reasonable.

**Comment by peixiu** [comment]
Test with the latest code:
PASS 1-smp_8.4096m.repeat1.run_test.Host_RHEL.m7.u5.raw.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_repetition.one_pci

File: qemu/tests/block_hotplug.py (hunk 4 above)
Comment (peixiu) [reply qemu/tests/block_hotplug.py]: Thanks coli, has updated the code, could you help to review it again?
thanks a lot~

**Comment by peixiu** [comment]
@CongLi @aliang123 could you help to  review it again? thanks a lot~

**Comment by aliang123** [comment]
Have tested with case: with_repetition(repeat time=20) + Win10.64, case passed.
 PASS 1-smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_repetition.one_pci 

Acked-by: Aihua Liang <aliang@redhat.com>

**Comment by CongLi** [comment]
Ack.

**Comment by peixiu** [comment]
@luckyh Hi xuhan, could you help to review this patch? thanks~

**Comment by peixiu** [comment]
There is a patch https://github.com/autotest/tp-qemu/pull/1370, it include a qmp event check after disk unpluged , it's more accurate than my settings before, so delete those codes, only keep add verify_unplug() code.

**Comment by peixiu** [comment]
Hi @CongLi @aliang123 could you help to review this patch again? I deleted some codes that have similar function with yhong's patch.
Thanks a lot~

**Comment by miyachen** [comment]
@peixiu there is a conflicting file, please take a look when you have time.

**Comment by peixiu** [comment]
Hi miya, thanks for your remind, currently, the bug that will be fixed by this patch has been closed as CURRENTRELEASE, it should be fixed by https://github.com/autotest/tp-qemu/pull/1396 together with https://github.com/avocado-framework/avocado-vt/pull/1558, so I close this patch now~

**Pull request closed without merging**